pkg/placement/placement-resolution.go

```diff
@@ -0,0 +1,363 @@
+// This function is thread-safe.
+func (resolution *placementResolution) updateObject(obj runtime.Object) (bool, error) {
+	resolution.Lock()
+	defer resolution.Unlock()
+
+	key, err := util.KeyForGroupVersionKindNamespaceName(obj)
+	if err != nil {
+		return false, fmt.Errorf("failed to get key for object: %w", err)
+	}
+
+	// formatted-key to use for mapping
+	formattedKey := key.GvkNamespacedNameKey()
+	_, exists := resolution.objectIdentifierToKeyMap[formattedKey]
+
+	// avoid further processing for keys of objects being deleted that do not have a deleted object
+	if isBeingDeleted(obj) || key.DeletedObject != nil {
+		delete(resolution.objectIdentifierToKeyMap, formattedKey)
+
+		return exists, nil
+	}
+
+	// add object to map
+	resolution.objectIdentifierToKeyMap[formattedKey] = &key
+	// if an object is reconciled all the way into here, it means something internal changed, so we need to trigger
+	// a placement-decision update event in order for transport watchers to re-evaluate it
+	return true, nil
+}
+
+// removeObject deletes an object from the resolution if it exists.
+// The return bool indicates whether the placement resolution was changed.
+// This function is thread-safe.
+func (resolution *placementResolution) removeObject(obj runtime.Object) bool {
+	resolution.Lock()
+	defer resolution.Unlock()
+
+	key, err := util.KeyForGroupVersionKindNamespaceName(obj)
+	if err != nil {
+		return false // assume object was never added
+	}
+
+	// formatted-key to use for mapping
+	formattedKey := key.GvkNamespacedNameKey()
+	_, exists := resolution.objectIdentifierToKeyMap[formattedKey]
+
+	delete(resolution.objectIdentifierToKeyMap, formattedKey)
+	return exists
+}
+
+// updateDestination updates the destinations list in the resolution.
+// This function is thread-safe.
+func (resolution *placementResolution) updateDestinations(destinations []string) {
+	resolution.Lock()
+	defer resolution.Unlock()
+
+	resolution.destinations = destinations
+}
+
+// toPlacementDecisionSpec converts the resolution to a placement decision
+// spec. This function is thread-safe.
+func (resolution *placementResolution) toPlacementDecisionSpec(gvkGvrMapper util.GvkGvrMapper) (*v1alpha1.PlacementDecisionSpec, error) {
+	resolution.RLock()
+	defer resolution.RUnlock()
+
+	workload := v1alpha1.DownsyncObjectReferences{}
+
+	// the following optimize the building of the workload by maintaining pointers to the object-wrapper structs
+	// that are key'd by GVR
+	ClusterScopeDownsyncObjectsMap := map[schema.GroupVersionResource]*v1alpha1.ClusterScopeDownsyncObjects{}
+	// Since namespaceScopeDownsyncObjectsMap groups objects by NS, these maps are used to efficiently locate the
+	// * struct by GVR
+	// * objects slice by namespace (for GVR)
+	namespaceScopeDownsyncObjectsMap := map[schema.GroupVersionResource]*v1alpha1.NamespaceScopeDownsyncObjects{}
+	nsObjectsLocationInSlice := map[string]int{} // key is GVR/ns to avoid 2D maps
```

Comment:
Why use a pointer (to a slice member) in one case and an index into the slice in the other case? Both work, I do not see a functional problem here; just wondering.

Reply:
Most likely a certain chain of thoughts at the time of writing this optimization led to this seeming more natural than using a pointer. I don't mind either.